### ISSUE TITLE
Pagination: clarify behavior of first/last calls. 

### ISSUE THREAD

**Issue by vslinko** [facebook/relay#735]
Copy of https://github.com/facebook/graphql/issues/104

I'm trying to implement pagination algorithm for my data and I don't know how to handle cases when `first` and `last` are negative.

What is your opinion about this?

ref: https://facebook.github.io/relay/graphql/connections.htm#sec-Pagination-algorithm

I think that spec should answer that question.

**Comment by josephsavona** [comment]
I agree with @leebyron's response in the original issue - `first` and `last` must be positive integers; this should be added to the Relay connection spec and `graphql-relay-js`'s connection helper should throw on invalid first/last values. 

Any interest in submitting a PR for this?

**Comment by vslinko** [comment]
I can submit PR for `graphql-relay-js` and I can try to submit to Relay spec, but I'm not sure about my English.

**Comment by BerndWessels** [comment]
Hi @josephsavona.

The specs say
> Any type whose name ends in “Connection” is considered by Relay to be a Connection Type.

Does that mean there is some magic in Relay around `connections` which would prevent from writing a custom connection?

I need to write a `connection` that can return the `current offset` with each page rather than only the `base64` encrypted `cursors`. Besides that I would also want the connection to refresh the `totalCount` with each page (in case it has changed). At the moment that only happens on the initial page.

Basically I want to have "real" pagination rather than the very very very limited `first after` `last before` pagination.

Unfortunately it seems that the following pagination is impossible with Relay at the moment:

![pagination](https://cloud.githubusercontent.com/assets/1776695/12314725/864507d0-bad8-11e5-9420-9ae4f40dc2e4.png)

**Comment by josephsavona** [comment]
@BerndWessels this issue is about the semenatics of first/last calls and is largely resolved. Since you've already opened another issue about pagination, can you ask this follow-up question on that issue instead?

**Comment by kassens** [comment]
Is there a reason not to allow 0? This might be useful for a case where you initially show no comments on an object and after pressing "see more comments" you increase from 0 upwards.

I would think we should update the spec and tests to non-negative. Thoughts?

**Comment by vslinko** [comment]
@kassens In current implementation `first: 0` means "give me them all".
I think it isn't good decision and I agree with you that `first: 0` could be helpful in some cases, so I can change my PRs.

@josephsavona what are you think?

**Comment by josephsavona** [comment]
I trust @kassens ;-)

My first thought was that `first:0` would be the same as `@include(if:false)`, but there might be fields on the connection that you want to fetch other than `edges`. Changing the semantics of first:0 to return zero edges makes sense, but we should verify that this won't impact products before making the change. Perhaps we could start with a warning in the writer if first is zero but edges are returned.

**Comment by KyleAMathews** [comment]
:+1: to `first: 0` just returning no edges. I'm an avid user of `connectionFields` and had been assuming I'd be able to use `first: 0` when I wanted no edge results.

**Comment by dschafer** [comment]
Facebook's GraphQL servers throw an error when it receives either zero or a negative number. Having negative numbers throw an error is clearly the right call. I can go either way on zero; I very much like the current Facebook GraphQL server behavior since sending `first: 0` is almost always an indication that the client didn't need to send that query, but I don't know if we need to express an error/empty opinion in the connection spec.

**Comment by wincent** [comment]
With 4eae620d86ed7fce1ee463c2fca88eb690d9, I believe there's nothing left to do here. Thanks to everybody for contributing to the discussion!
